_posts/2024-04-29-dependabot-management.markdown

```diff
@@ -0,0 +1,206 @@
+
+### AWS CodeArtifact and GitHub Dependabot Configuration
+
+Unfortunately, to solve this issue it is not super straight forward and numerous poor tutorials exist online.
+To use Dependabot with AWS CodeArtifact, a registry needs to be defined in the `.github/dependabot.yml` file for each repository. This is unfortunately not supported at an organization level and uses a token which can be defined in the organization secrets. Though there are some important security considerations for that token.
+
+Here is an example of the dependabot configuration for a repository using AWS CodeArtifact.
+
+```yaml
+version: 2
+registries:
+  npm-codeartifact:
+    type: npm-registry
+    url: https://**ANONYMOUS**.d.codeartifact.ap-southeast-2.amazonaws.com/npm/DefaultCodeArtifactRepository-**ANONYMOUS**/
+    token: ${{ secrets.AWS_CODEARTIFACT_TOKEN }}
+
+updates:
+  - package-ecosystem: "npm"
+    registries:
+      - npm-codeartifact
+    directory: "/"
+    #  ... see code example above for the rest of the configuration
+```
+
+Above, the registry `npm-codeartifact` is defined with the URL of the AWS CodeArtifact repository and the token is defined in the organization secrets.
+
+Most tutorial's recommend using a long lived AWS token for this. However, for a organization using Github, it is more appropriate to create or use a separate repository for provisioning organization level resources. In the case of AWS CodeArtifact, within this separate repository create an OIDC connection between this repo and AWS. This way there is a secure connection to AWS inside a Github Action's workflow.
+
+Then create a workflow to automate the creation of short lived tokens for Dependabot that are stored as organization level secrets. This way the token is only valid for a short period of time and is rotated regularly. This is a more secure way of managing the token and is in line with the principle of least privilege.
+
+```yaml
+name: Manage AWS CodeArtifact Secret
+
+# This GitHub workflow, automates the process of creating, or updating a secret at the organization level using a GitHub App.
+
+on:
+  push:
+    branches:
+      - main
+  schedule:
+    - cron: "50 */7 * * *" # At minute 50 past every 7th hour every day.
+
+jobs:
+  manage-secret:
+   #  ...
+    steps:
+      - uses: actions/checkout@v3
+      #...
+      - name: Set AWS CodeArtifact Secrets for Actions and Dependabot
+        env:
+          GITHUB_TOKEN: ${{ steps.generate_token.outputs.token }}
+          ORG: *YOUR_ORG_NAME*
+        run: |
+          aws codeartifact get-authorization-token --domain *ANONYMOUS* --domain-owner *ANONYMOUS* --query authorizationToken --output text --region ap-southeast-2 --duration-seconds 28800 > aws-codeartifact-token.txt
+
+          gh secret set AWS_CODEARTIFACT_TOKEN --app dependabot --org ${{ env.ORG }} --visibility all < aws-codeartifact-token.txt
+```
+
+<details>
+<summary>
+SEE FULL CODE HERE
+</summary>
+
+```yaml
```

Comment:
Is this doubled up from above?

Reply:
one is full code, one is abbreviated in summary of detail 
https://gist.github.com/scmx/eca72d44afee0113ceb0349dd54a84a2

Reply:
kinda ...